arenadata/db
# Beforehand generation of transaction ID when inserting into AO tables.

Empty transaction ID can cause errors on dependent entities calculation, such as spgist indexes.
Adding a call to 'GetCurrentTransactionId' like it's done in vanilla Postgres version of 'heap_insert' function.
Each of 4 regression tests is independent.
PR comment and commit message will be updated after discussion.
______________________
QA
- Why not just generate new transaction ID [here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/executor/nodeModifyTable.c#L479)?
- There are other operators affected. `UPDATE`, `COPY` and, possibly, `ALTER TABLE`.
__
- ~~Why no `UPDATE` regression tests?~~
- ~~AO table update logic reuses insert logic([one](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/access/appendonly/appendonlyam.c#L2576), [two](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/access/aocs/aocsam.c#L1625)).~~
__
- Why no `ALTER TABLE` regression tests?
- I was not able to find working solution to trigger an error inside `ATRewriteTable` [function](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/commands/tablecmds.c#L6274). As I can see, transaction ID correctly generated on previous steps for all cases I found.
__
- Why not place a fix [here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/executor/nodeModifyTable.c#L564), [here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/executor/execUtils.c#L1129) or even [here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/executor/execUtils.c#L1198)?
- It's also possible to place a fix in AO init functions([here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/access/appendonly/appendonlyam.c#L2595) and [here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/access/aocs/aocsam.c#L824)). But it seems to be implicitly logical  to place generation of new transaction ID into `insert` function, and more, this solution [inherits](https://github.com/postgres/postgres/blob/REL9_4_STABLE/src/backend/access/heap/heapam.c#L2062) the way of vanilla Postgres version of function `heap_insert`. The drawback of this solution, it triggers the call to `GetCurrentTransactionId` even for non-affected entities, such as [matview](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/commands/matview.c#L575) and [ctas](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/commands/createas.c#L714). At this time, I see it may be preferable to place a fix somewhere inside `ExecInsertIndexTuples`, despite the fact that it contradicts vanilla Postgres way. I want to hear teams opinion.
